Fixes #25577 - updated puppetclass_lookup_keys_controller queries

ShimShtein opened the pull request:
1. Added `:inverse_of` to relevant objects
2. Changed queries to use less resources.

mmoll commented:
[test katello]

tbrisker commented on app/controllers/puppetclass_lookup_keys_controller.rb:
```diff
@@ -4,8 +4,9 @@ class PuppetclassLookupKeysController < LookupKeysController
 
   def index
-    @lookup_keys = resource_base_search_and_page(:param_classes).smart_class_parameters
-    @puppetclass_authorizer = Authorizer.new(User.current, :collection => @lookup_keys.map {|key| key.param_class.try(:id)}.compact.uniq)
```
Why did you remove this? afaict it is still used in the view

ShimShtein replied:
By mistake. Now it's back

tbrisker commented on app/controllers/puppetclass_lookup_keys_controller.rb:
```diff
@@ -4,8 +4,9 @@ class PuppetclassLookupKeysController < LookupKeysController
-    @puppetclass_authorizer = Authorizer.new(User.current, :collection => @lookup_keys.map {|key| key.param_class.try(:id)}.compact.uniq)
+    @lookup_keys = resource_base_search_and_page.distinct.smart_class_parameters.preload(:lookup_values)
+    environment_classes = EnvironmentClass.where(puppetclass_lookup_key_id: @lookup_keys.map(&:id)).select(:puppetclass_id, :puppetclass_lookup_key_id).distinct.preload(:puppetclass)
+    @lookup_keys_to_class = Hash[environment_classes.map { |environment_class| [environment_class.puppetclass_lookup_key_id, environment_class.puppetclass]}]
```
`.smart_class_parameters` already joins environment_class, so I don't think we need to toad the environment classes again - we should already have access to `environment_class.puppetclass_id` in `@lookupkeys`

ShimShtein replied:
`.smart_class_parameters` supplies the `.joins` statement, but it does not add the environment_classes relation to the results. (We can't add it to the result, since it will mess up the pagination).
The new code preloads lookup_values with lookup keys and preloads puppetclasses with stripped down environment classes. I need the environment_class object for the lookup key connection.

tbrisker reviewed (changes requested):
A couple of comments inline, can you provide some before/after numbers for this change?

ShimShtein commented:
@tbrisker, updated the PR, can you review again?

tbrisker commented on app/controllers/puppetclass_lookup_keys_controller.rb:
```diff
@@ -4,8 +4,11 @@ class PuppetclassLookupKeysController < LookupKeysController
   def index
-    @lookup_keys = resource_base_search_and_page(:param_classes).smart_class_parameters
-    @puppetclass_authorizer = Authorizer.new(User.current, :collection => @lookup_keys.map {|key| key.param_class.try(:id)}.compact.uniq)
+    @lookup_keys = resource_base_search_and_page.distinct.smart_class_parameters.preload(:lookup_values)
```
Since we are going directly to EnvironmentClass, why do we need `smart_class_parameters` here?

tbrisker commented on app/controllers/puppetclass_lookup_keys_controller.rb:
```diff
@@ -4,8 +4,11 @@ class PuppetclassLookupKeysController < LookupKeysController
-    @lookup_keys = resource_base_search_and_page(:param_classes).smart_class_parameters
-    @puppetclass_authorizer = Authorizer.new(User.current, :collection => @lookup_keys.map {|key| key.param_class.try(:id)}.compact.uniq)
+    @lookup_keys = resource_base_search_and_page.distinct.smart_class_parameters.preload(:lookup_values)
+    environment_classes = EnvironmentClass.where(puppetclass_lookup_key_id: @lookup_keys.map(&:id)).select(:puppetclass_id, :puppetclass_lookup_key_id).distinct.preload(:puppetclass)
```
Perhaps add index on environment_class(:puppetclass_lookup_key_id, :puppetclass_id)

ShimShtein commented:
@tbrisker, addressed both your comments

tbrisker commented:
looks like the failing test is related, but i am not sure it's a valid test anymore. @orrabin, iirc we decided to remove orphaned class params a while ago?
Edit: yes, we did, in https://github.com/theforeman/foreman/blob/develop/db/migrate/20161205142618_delete_orphaned_smart_class_parameters.rb

ShimShtein committed:
Fixes #25577 - updated puppetclass_lookup_keys_controller queries

1. Added :inverse_of to relevant objects
2. Changed queries to use less resources.

ShimShtein commented:
@tbrisker finally the tests are passing. Can you review it again?

tbrisker reviewed (approved):
Thanks @ShimShtein !
